crates/bevy_render/src/view/mod.rs

```diff
@@ -525,3 +534,73 @@ fn prepare_view_targets(
         }
     }
 }
+
+#[derive(PipelineKeyInRenderCrate, Default, Clone, Copy, PartialEq, Eq, Hash, Debug)]
+#[repr(u8)]
+#[custom_shader_defs]
+pub enum MsaaKey {
+    #[default]
+    Off,
+    X2,
+    X4,
+    X8,
+    X16,
+    X32,
+    X64,
+    X128,
+}
+impl SystemKey for MsaaKey {
+    type Param = SRes<Msaa>;
+    type Query = ();
+
+    fn from_params(msaa: &Res<Msaa>, _: ()) -> Option<Self> {
+        Some((msaa.samples().trailing_zeros() as u8).into())
+    }
+}
+
+impl KeyShaderDefs for MsaaKey {
+    fn shader_defs(&self) -> Vec<ShaderDefVal> {
+        let samples = ShaderDefVal::UInt("MSAA_SAMPLECOUNT".to_owned(), self.samples());
+        match self {
+            MsaaKey::Off => vec![samples],
+            _ => vec!["MULTISAMPLED".into(), samples],
+        }
+    }
+}
+
+impl MsaaKey {
+    pub fn samples(&self) -> u32 {
+        1 << (*self as u32)
+    }
+}
+
+#[derive(PipelineKeyInRenderCrate, Clone, Copy, PartialEq, Eq, Hash, Debug)]
+#[custom_shader_defs]
+pub struct TextureFormatKey(pub bool);
+impl SystemKey for TextureFormatKey {
+    type Param = ();
+    type Query = Read<ExtractedView>;
+
+    fn from_params(_: &(), view: &ExtractedView) -> Option<Self> {
+        Some(Self(view.hdr))
+    }
+}
+
+impl KeyShaderDefs for TextureFormatKey {
+    fn shader_defs(&self) -> Vec<ShaderDefVal> {
+        match self.0 {
+            false => vec!["TONEMAP_IN_SHADER".into()],
+            true => vec!["HDR".into()],
+        }
```

Comment:
Could just be `if` instead of `match`

Reply:
how do you feel about `self.0.then_some("TONEMAP_IN_SHADER".into()).into_iter().collect()` ?

i did that on `GizmoConfigKey` for example. it should definitely be consistent whichever one we use though.